client/views/view_units.cpp

```diff
@@ -262,57 +282,66 @@ void units_view::update_view()
  */
 void units_view::update_waiting()
 {
-  int units_count = 0;
+  QTableWidgetItem *item;
+  QFont f = QApplication::font();
+  QFontMetrics fm(f);
+  int h = fm.height() + 10;
 
+  ui.uwt_widget->setRowCount(0);
   ui.uwt_widget->clearContents();
 
-  if (!client_has_player()) {
-    return;
-  }
+  struct unit_waiting_entry unit_entries[U_LAST];
+  int entries_used = 0;
 
-  unit_list_iterate(client_player()->units, punit)
-  {
-    if (!can_unit_move_now(punit) && punit->ssa_controller == SSA_NONE) {
-      QTableWidgetItem *item =
-          new QTableWidgetItem(utype_name_translation(punit->utype));
-      item->setData(Qt::UserRole,
-                    QVariant::fromValue(static_cast<void *>(punit)));
-      ui.uwt_widget->setItem(units_count, 0, item);
-
-      int pcity_near_dist;
-      struct city *pcity_near = get_nearest_city(punit, &pcity_near_dist);
-      ui.uwt_widget->setItem(
-          units_count, 1,
-          new QTableWidgetItem(
-              get_nearest_city_text(pcity_near, pcity_near_dist),
-              pcity_near_dist));
-
-      ui.uwt_widget->setItem(
-          units_count, 2,
-          new QTableWidgetItem(move_points_text(punit->moves_left, false),
-                               punit->moves_left));
-
-      time_t dt = time(nullptr) - punit->action_timestamp;
-      if (dt < 0 && !can_unit_move_now(punit)) {
-        char buf[64];
-        format_time_duration(-dt, buf, sizeof(buf));
-        ui.uwt_widget->setItem(units_count, 3,
-                               new QTableWidgetItem(buf, dt));
-      }
+  get_units_waiting_data(unit_entries, &entries_used);
+
+  max_row = 0;
+  int i, j;
+  for (i = 0; i < entries_used; i++) {
+    struct unit_waiting_entry *pentry = unit_entries + i;
+    struct unit_type *putype = pentry->type;
+    cid id;
+
+    auto sprite = get_unittype_sprite(tileset, putype, direction8_invalid());
+    id = cid_encode_unit(putype);
 
-      ++units_count;
+    ui.uwt_widget->insertRow(max_row);
+    for (j = 0; j < 3; j++) {
+      item = new QTableWidgetItem;
```

Comment:
Declare when first used
```suggestion
      auto item = new QTableWidgetItem;
```